### ISSUE TITLE
How to implement removed CC_GCC_EXTRA_ENV_ARRAY?

### ISSUE THREAD

**Issue by jasmin-j** [crosstool-ng/crosstool-ng#323]
See #313 (remove CC_GCC_EXTRA_ENV_ARRAY)

I am currently on vacation in the Philippines and just realized that you removed my extension.
Now my arm-none-eabi cross compiler will no longer build :(
The ARM build script uses this to circumvent a problem (I forget the details), when building the gcc target libraries. So using " INHIBIT_LIBC_CFLAGS='-DUSE_TM_CLONE_REGISTRY=0' " IS the solution from ARM.
I was really happy with CT-NG and tried to focus on the Python GDB problem on MINGW for the arm-none-eabi Windows version. Now I am 2 months back ... .

@bhundven:
Do you really mean we should add for any compiler config flags, which may even version dependent?
Shall I really add a config option to set the INHIBIT_LIBC_CFLAGS ?
My old solution could be used to add ANY flag to the make command line, which may be useful to solve all sorts of future problems, too.
What is your proposed solution to add this flag to the build?

BR
Jasmin

**Comment by bhundven** [comment]
@jasmin-j This ```-DUSE_TM_CLONE_REGISTRY``` seems to be arm specific, but I'm showing my ignorance of this option. I'm not sure I understand why it is required.

Help me to understand the circumstance it is needed will help us to add this correctly.

**Comment by jasmin-j** [comment]
I found this in the ARM build script:

~~~
# Passing USE_TM_CLONE_REGISTRY=0 via INHIBIT_LIBC_CFLAGS to disable
# transactional memory related code in crtbegin.o.
# This is a workaround. Better approach is have a t-* to set this flag via
# CRTSTUFF_T_CFLAGS
if [ "x$DEBUG_BUILD_OPTIONS" != "x" ]; then
  make -j$JOBS CXXFLAGS="$DEBUG_BUILD_OPTIONS" \
           INHIBIT_LIBC_CFLAGS="-DUSE_TM_CLONE_REGISTRY=0"
~~~

In file "libgcc/crtstuff.c" you can grep for USE_TM_CLONE_REGISTRY. If it is set to "0", the transactional
memory initialization code is not emitted.
The default of that flag is "1". I have no clue why ARM does disable it for the embedded processors, but it
seems necessary.

**Comment by bhundven** [comment]
I guess my questions are:

* How did you find out you needed this?
* Is it needed for all arm targets, or just specific arm targets?
* Are there other targets that require this? I notice that Alpha specifically sets this to 0.

**Comment by jasmin-j** [comment]
> How did you find out you needed this?

As I wrote in my post, ARM uses it in the building process for the embedded processors.

> Is it needed for all arm targets, or just specific arm targets?

For all embedded targets, like Cortex-M0, M1, .... .

> Are there other targets that require this? I notice that Alpha specifically sets this to 0.

Yes, they are using another approach to set it. In my opinion the right approach.

I am still in the Philippines with very bad internet, but as soon I come back, I will  continue my work
on the arm-none-eabi toolchain. There is now GCC 5.3.0, Binutils 2.26 and, GDB 7.10.1 officially released. Moreover newlib is available in snapshots now, so all fixes/extensions for arm-none-eabi should be available also (snapshot support and patches for them needs to be added to CT-NG!).

I guess, I can build the toolchain from the official sources, with only a few patches.
When I do this, I will look how I can add this flag during the multilib build process for only this embedded targets, like the Alpha does it. An no, I didn't have analysed why ARM is disabling the transactional memory initialization code for the embedded toolchain. I accepted, that they did it!
Moreover, I don't know if others like to have the Cortex-Mx toolchain built with the flag set. If not, we need target depending patches which is currently not supported by CT-NG.

**Comment by bhundven** [comment]
> How did you find out you needed this?
> 
> As I wrote in my post, ARM uses it in the building process for the embedded processors.
> 
I mean, did something fail in a way that it shouldn’t have? Is there an open issue on the gcc bugzilla?
> Is it needed for all arm targets, or just specific arm targets?
> 
> For all embedded targets, like Cortex-M0, M1, .... .
> 
So it is needed for No MMU micro-controllers only?

> Are there other targets that require this? I notice that Alpha specifically sets this to 0.
> 
> Yes, they are using another approach to set it. In my opinion the right approach.
> 
> I am still in the Philippines with very bad internet, but as soon I come back, I will continue my work
> on the arm-none-eabi toolchain.
> 
Ok!
> There is now GCC 5.3.0, Binutils 2.26 and, GDB 7.10.1 officially released.
> 
Yup, and they are now in crosstool-ng master branch! :)
> Moreover newlib is available in snapshots now, so all fixes/extensions for arm-none-eabi should be available also (snapshot support and patches for them needs to be added to CT-NG!).
> 
Yes, I have them in PR #320, but I need someone to review them. I’ve been waiting for a review on them for a while now.
> I guess, I can build the toolchain from the official sources, with only a few patches.
> 
Which patches? I thought we were just adding a flag?
> When I do this, I will look how I can add this flag during the multilib build process for only this embedded targets, like the Alpha does it. An no, I didn't have analysed why ARM is disabling the transactional memory initialization code for the embedded toolchain. I accepted, that they did it!
> 
Yea, I would like to understand why we do things in crosstool-ng instead of just passing random arguments to gcc and hoping for the best. Which is why I removed GCC_EXTRA_ENV_ARRAY in the first place.
> Moreover, I don't know if others like to have the Cortex-Mx toolchain built with the flag set. If not, we need target depending patches which is currently not supported by CT-NG.
> 
Sounds complex.

**Comment by jasmin-j** [comment]
> I mean, did something fail in a way that it shouldn’t have? Is there an open issue on the gcc bugzilla?

As I wrote in my first post, I try to use CT-NG to build an arm-none-eabi toolchain, which is built as close as possible to the one provided by ARM (see first post). They disabled the transactional memory initialization code for any reason. The target for me is to have the toolchain as close as possible to the one from ARM and not to check why they did this or that.

I am not sure if ARM will answer me why they disabled it, but if you like I can try to ask them. I have a contact eMail from a developer there.

> So it is needed for No MMU micro-controllers only?

The arm-none-eabi toolchain is for the Cortex-Mx controllers only and there they added this. They even have their own branch for this processors. On that branch they recently added the armv8-m target, too. I haven't analyzed this new architecture, but as far as I remember some changelog entries, this type has security features implemented also. I currently can't check, if it has an MMU.
So I can't answer your question, to be honest.

> Yes, I have them in PR #320, but I need someone to review them. I’ve been waiting for a review on them for a while now.

I can do that, when I am back and start to use the newest newlib variant.

> Which patches? I thought we were just adding a flag?

To build the arm-none-eabi toolchain, I need to enable multilib compilation for all the Cortex-Mx processors. I tried to release the patch for GCC, but it was too late for 5.3.0. Moreover ARM is changed the original patch and a small fix from me again. The GCC maintainers are sometimes really complicated and this is frustrating. Adding it as a patch to CT-NG will not harm anything, but extend it simply.

>> Moreover, I don't know if others like to have the Cortex-Mx toolchain built with the flag set. If not, we need target depending patches which is currently not supported by CT-NG.

> Sounds complex.

But easy to solve with a new configure option to GCC, which can then be added by the custom flags per toolchain variant.
